### PR TITLE
[Perf] [Model] Fuse RoPE invocations in QwenImageCrossAttention

### PR DESCRIPTION
**Pull request by kliuae**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

This PR reduces rotary embedding overhead in QwenImageCrossAttention by fusing previously separate RoPE calls into one.

Instead of applying RoPE independently for query and key tensors of image and text segments, this PR concatenates the q/k tensors, leveraging the fact that txt and img are combined along the sequence dimension, before feeding the combined tensor into RoPE, so that it only needs to be executed once.

Benchmarking on Qwen-Image series, it shows improvement to end-to-end latency of ~11% for Qwen-Image-Edit-2511 and ~29% for Qwen-Image-2512 on AMD MI300X.

## Test Plan

Performance benchmark on MI300X

## Test Result

Qwen/Qwen-Image-2512 (vbench; t2i; max concurrency=1; 5 requests)
Metric | Before | This PR
-- | -- | --
Benchmark duration (s) | 78.12 | 55.13
Request throughput (req/s) | 0.06 | 0.09
Latency mean (s) | 15.6236 | 11.0263
Latency median (s) | 15.6525 | 11.0563
Latency p99 (s) | 15.6837 | 11.0821

Qwen/Qwen-Image-Edit-2511 (vbench; i2i; max concurrency=1; 5 requests)
Metric | Before | This PR
-- | -- | --
Benchmark duration (s) | 139.95 | 123.91
Request throughput (req/s) | 0.04 | 0.04
Latency mean (s) | 27.9895 | 24.7815
Latency median (s) | 28.0357 | 24.8072
Latency p99 (s) | 28.0508 | 24.8454

### Generation test
Model: Qwen/Qwen-Image-2512
Prompt: "a cup of coffee on the table"

Before:
base2_output_image_text2image
This PR:
rope_concat2_output_image_text2image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by ZJY0516** [comment]
I'm noticing a small performance drop on gpu h20

```
python diffusion_benchmark_serving.py --port 8000 --model Qwen/Qwen-Image-2512 --task t2i --dataset random --num-prompts 5
```

main
```
================= Serving Benchmark Result =================
Model:                                   Qwen/Qwen-Image-2512
Dataset:                                 random         
Task:                                    t2i            
--------------------------------------------------
Benchmark duration (s):                  140.65         
Request rate:                            inf            
Max request concurrency:                 1              
Successful requests:                     5/5              
--------------------------------------------------
Request throughput (req/s):              0.04           
Latency Mean (s):                        28.1293        
Latency Median (s):                      28.0871        
Latency P99 (s):                         28.3050        

============================================================
```

this pr

```
================= Serving Benchmark Result =================
Model:                                   Qwen/Qwen-Image-2512
Dataset:                                 random         
Task:                                    t2i            
--------------------------------------------------
Benchmark duration (s):                  141.54         
Request rate:                            inf            
Max request concurrency:                 1              
Successful requests:                     5/5              
--------------------------------------------------
Request throughput (req/s):              0.04           
Latency Mean (s):                        28.3084        
Latency Median (s):                      28.2591        
Latency P99 (s):                         28.4509        

============================================================
```

File: vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py [hunk 1]
```diff
@@ -534,16 +534,19 @@ def forward(
...
         joint_key = torch.cat([txt_key, img_key], dim=1)
         joint_value = torch.cat([txt_value, img_value], dim=1)
 
+        # This leverages the same dtype used by txt_qkv and img_qkv,
+        # since the value tensors can be concatenated.
+        joint_qk = torch.cat([joint_query, joint_key], dim=0)
+        joint_cos = torch.cat([txt_cos, img_cos], dim=0)
+        joint_sin = torch.cat([txt_sin, img_sin], dim=0)
+        joint_qk = self.rope(joint_qk, joint_cos, joint_sin)
+        joint_query, joint_key = torch.chunk(joint_qk, 2, dim=0)
```
[the remark this answers is not in the document]
Comment (kliuae) [reply vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py:548]: Addressed. Split for the SP branch

**Comment by hsliuustc0106** [comment]
> I'm noticing a small performance drop on gpu h20
> 
> ```
> python diffusion_benchmark_serving.py --port 8000 --model Qwen/Qwen-Image-2512 --task t2i --dataset random --num-prompts 5
> ```
> 
> main
> 
> ```
> ================= Serving Benchmark Result =================
> Model:                                   Qwen/Qwen-Image-2512
> Dataset:                                 random         
> Task:                                    t2i            
> --------------------------------------------------
> Benchmark duration (s):                  140.65         
> Request rate:                            inf            
> Max request concurrency:                 1              
> Successful requests:                     5/5              
> --------------------------------------------------
> Request throughput (req/s):              0.04           
> Latency Mean (s):                        28.1293        
> Latency Median (s):                      28.0871        
> Latency P99 (s):                         28.3050        
> 
> ============================================================
> ```
> 
> this pr
> 
> ```
> ================= Serving Benchmark Result =================
> Model:                                   Qwen/Qwen-Image-2512
> Dataset:                                 random         
> Task:                                    t2i            
> --------------------------------------------------
> Benchmark duration (s):                  141.54         
> Request rate:                            inf            
> Max request concurrency:                 1              
> Successful requests:                     5/5              
> --------------------------------------------------
> Request throughput (req/s):              0.04           
> Latency Mean (s):                        28.3084        
> Latency Median (s):                      28.2591        
> Latency P99 (s):                         28.4509        
> 
> ============================================================
> ```

this looks like zero difference

**Comment by ZJY0516** [comment]
> this looks like zero difference

No, it's always slightly slow than main

**Comment by ZJY0516** [comment]
same performance drop on H200

main

```
================= Serving Benchmark Result =================
Model:                                   /data/datasets/models-hf/Qwen-Image-2512/
Dataset:                                 random         
Task:                                    t2i            
--------------------------------------------------
Benchmark duration (s):                  41.45          
Request rate:                            inf            
Max request concurrency:                 1              
Successful requests:                     5/5              
--------------------------------------------------
Request throughput (req/s):              0.12           
Latency Mean (s):                        8.2899         
Latency Median (s):                      8.3071         
Latency P99 (s):                         8.3687         

============================================================
```


this pr
```
================= Serving Benchmark Result =================
Model:                                   /data/datasets/models-hf/Qwen-Image-2512/
Dataset:                                 random         
Task:                                    t2i            
--------------------------------------------------
Benchmark duration (s):                  41.91          
Request rate:                            inf            
Max request concurrency:                 1              
Successful requests:                     5/5              
--------------------------------------------------
Request throughput (req/s):              0.12           
Latency Mean (s):                        8.3825         
Latency Median (s):                      8.3895         
Latency P99 (s):                         8.4551         

============================================================
```

**Comment by kliuae** [comment]
> same performance drop on H200
> 
> main
> 
> ```
> ================= Serving Benchmark Result =================
> Model:                                   /data/datasets/models-hf/Qwen-Image-2512/
> Dataset:                                 random         
> Task:                                    t2i            
> --------------------------------------------------
> Benchmark duration (s):                  41.45          
> Request rate:                            inf            
> Max request concurrency:                 1              
> Successful requests:                     5/5              
> --------------------------------------------------
> Request throughput (req/s):              0.12           
> Latency Mean (s):                        8.2899         
> Latency Median (s):                      8.3071         
> Latency P99 (s):                         8.3687         
> 
> ============================================================
> ```
> 
> this pr
> 
> ```
> ================= Serving Benchmark Result =================
> Model:                                   /data/datasets/models-hf/Qwen-Image-2512/
> Dataset:                                 random         
> Task:                                    t2i            
> --------------------------------------------------
> Benchmark duration (s):                  41.91          
> Request rate:                            inf            
> Max request concurrency:                 1              
> Successful requests:                     5/5              
> --------------------------------------------------
> Request throughput (req/s):              0.12           
> Latency Mean (s):                        8.3825         
> Latency Median (s):                      8.3895         
> Latency P99 (s):                         8.4551         
> 
> ============================================================
> ```

Thanks for helping validate this. I don't have a CUDA environment at hand to collect the benchmarks.
Given the results, I'll restrict the changes to ROCm for now.

**Review by gcanlin — COMMENTED** [review]
On NPU(A3), it brings a little improvement from 50.8s -> 49.4s for Qwen-Image without any parallelism(based on v0.12.0rc1). I think it's a general optimization even if it's little.

**Comment by tjtanaa** [comment]
@ZJY0516 @hsliuustc0106 should we include this into milestone?

**Comment by lishunyang12** [comment]
@kliuae Hey, fusing the separate RoPE calls for q/k tensors into one is a clean optimization — just 12 lines changed. Did you get benchmark numbers for the latency improvement on Qwen-Image? This seems straightforward to merge if the numbers check out.
